File: src/graphics_view.py
# Copyright (C) 2020-2022 viraelin
# License: GPLv3.0

import os
import math
import tempfile
import rpack
import shutil
import urllib.request
import time

from PIL import (Image, UnidentifiedImageError)
from PyQt6.QtCore import *
from PyQt6.QtWidgets import *
from PyQt6.QtGui import *

import system
import commands
from graphics_item import GraphicsItem


class GraphicsView(QGraphicsView):

    itemMoved = pyqtSignal(list, list)

    def __init__(self, scene: QGraphicsScene, parent=None) -> None:
        super().__init__(scene, parent=parent)

        scene.selectionChanged.connect(self.onSelectionChanged)
        self._mouse_last_press_position = QPointF()
        self._mouse_last_drop_position = QPointF()

        self.transformation_mode = Qt.TransformationMode.SmoothTransformation
        self.grayscale_effect = QGraphicsColorizeEffect()
        self.grayscale_effect.setColor(0)
        self.grayscale_effect.setEnabled(False)
        self.setGraphicsEffect(self.grayscale_effect)
        self.setHorizontalScrollBarPolicy(Qt.ScrollBarPolicy.ScrollBarAlwaysOff)
        self.setVerticalScrollBarPolicy(Qt.ScrollBarPolicy.ScrollBarAlwaysOff)
        self.setAcceptDrops(True)
        self.setDragMode(QGraphicsView.DragMode.RubberBandDrag)
        self.setStyleSheet("background-color: #100111111;")


    def load(self):
        image_data = system.sql.loadImages()
        for entry in image_data:
            item_id = entry["id"]

            item_x = entry["x"]
            item_y = entry["y"]
            item_z_value = entry["z"]
            item_rotation = entry["rotation"]
            item_scale = entry["scale"]
            item_is_flipped = entry["flip"]
            item_image = entry["image"]

            pixmap = QPixmap()
            pixmap.loadFromData(item_image)

            item = GraphicsItem(item_id, pixmap)

            item.path = entry["path"]
            item.type = entry["type"]
            item.ctime = entry["ctime"]
            item.mtime = entry["mtime"]

            item.setPos(item_x, item_y)
            item.setZValue(item_z_value)
            item.setScale(item_scale)
            item.setRotation(math.degrees(item_rotation))
            item.setTransformationMode(self.transformation_mode)
            if item_is_flipped:
                item.flip()
            self.scene().addItem(item)
            system.item_ids.append(item_id)

        view_data = system.sql.loadView()
        view_x = view_data["x"]
        view_y = view_data["y"]
        view_scale = view_data["scale"]
        view_pos = QPointF(view_x, view_y)

        self.setTransformationAnchor(QGraphicsView.ViewportAnchor.AnchorViewCenter)
        self.updateSceneSize()
        default_scale = 1.0/self.transform().m11()
        self.scale(default_scale, default_scale)
        self.scale(view_scale, view_scale)
        self.centerOn(view_pos)


    def mousePressEvent(self, event: QMouseEvent) -> None:
        if event.button() == Qt.MouseButton.LeftButton:
            self._mouse_last_press_position = event.position()
            if event.modifiers() == Qt.KeyboardModifier.ShiftModifier:
                event.accept()
                return
            if event.modifiers() == Qt.KeyboardModifier.ControlModifier:
                event.accept()
                return
        if event.button() == Qt.MouseButton.RightButton:
            # context menu
            return
        if event.button() == Qt.MouseButton.MiddleButton:
            self._mouse_last_press_position = event.position()
            event.accept()
            return
        super().mousePressEvent(event)


    def mouseReleaseEvent(self, event: QMouseEvent) -> None:
        super().mouseReleaseEvent(event)
        if event.button() == Qt.MouseButton.LeftButton:
            items = self.scene().selectedItems()
            if len(items) > 0:
                p1 = self.mapToScene(self._mouse_last_press_position.toPoint())
                p2 = self.mapToScene(event.position().toPoint())
                offset = p2 - p1
                if offset:
                    system.undo_stack.push(commands.Move(items, offset, True))


    def mouseMoveEvent(self, event: QMouseEvent) -> None:
        if event.buttons() == Qt.MouseButton.LeftButton:
            if event.modifiers() == Qt.KeyboardModifier.ShiftModifier:
                self.rotateSelection(event)
                event.accept()
                return
            if event.modifiers() == Qt.KeyboardModifier.ControlModifier:
                self.scaleSelection(event)
                event.accept()
                return
        if event.buttons() == Qt.MouseButton.MiddleButton:
            self.panView(event)
            event.accept()
            return
        super().mouseMoveEvent(event)


    def wheelEvent(self, event: QWheelEvent) -> None:
        self.zoomView(event)


    def dragMoveEvent(self, event: QDragMoveEvent) -> None:
        event.acceptProposedAction()


    def dragEnterEvent(self, event: QDragEnterEvent) -> None:
        mimedata = event.mimeData()
        if mimedata.hasUrls():
            event.acceptProposedAction()
        elif mimedata.hasImage():
            event.acceptProposedAction()


    def dropEvent(self, event: QDropEvent) -> None:
        mimedata = event.mimeData()
        pos = self.mapToScene(event.position().toPoint())
        self._mouse_last_drop_position = pos

        can_store = False
        if system.sql.file_path != "":
            can_store = True

        items = []
        if mimedata.hasUrls():
            urls = mimedata.urls()
            print("urls: ", urls)
            for url in urls:
                if url.isLocalFile():
                    path = url.path()
                    item = self.createItem(path, pos, can_store=can_store)
                    items.append(item)
                else:
                    s_url = url.url()
                    with urllib.request.urlopen(s_url) as response:
                        with tempfile.NamedTemporaryFile() as temp_file:
                            shutil.copyfileobj(response, temp_file)
                            path = temp_file.name
                            item = self.createItem(path, pos, url=s_url, can_store=can_store)
                            items.append(item)
        elif mimedata.hasImage():
            print("image: ")

        if len(items) > 1:
            self.scene().clearSelection()
            for item in items:
                item.setSelected(True)
            self.packSelection()


    def createItem(self, path: str, pos: QPointF, is_flipped=False, scale=1.0, z_value=0, url="", can_store=False) -> GraphicsItem:
        file_type = None
        try:
            img = Image.open(path)
            file_type = img.format
        except UnidentifiedImageError:
            return None

        basename = os.path.basename(path)
        if url != "":
            basename = os.path.basename(url)

        name, ext = os.path.splitext(basename)

        ext_changed = False
        if ext == "":
            ext = "." + file_type.lower()
            if ext == ".jpeg":
                ext = ".jpg"
            ext_changed = True

        if name == ext:
            with tempfile.NamedTemporaryFile() as temp_file:
                basename = os.path.basename(temp_file.name)

        if ext_changed:
            basename += ext

        new_id = system.getItemID()

        item = GraphicsItem(new_id, QPixmap(path))

        item.path = basename
        item.mtime = os.path.getmtime(path)
        item.ctime = time.time()

        item.type = file_type
        item.setPos(pos)
        item.setScale(scale)
        item.setZValue(z_value)
        item.setTransformationMode(self.transformation_mode)

        if is_flipped:
            item.flip()

        self.scene().addItem(item)
        center = item.boundingRect().center()
        item.setPos(item.pos() - center)

        if can_store:
            system.sql.storeItem(item)

        return item


    def panView(self, event: QMouseEvent) -> None:
        if not len(self.scene().items()) > 0:
            return

        self.setTransformationAnchor(QGraphicsView.ViewportAnchor.NoAnchor)
        self.updateSceneSize()
        old_pos = self.mapToScene(self._mouse_last_press_position.toPoint())
        new_pos = self.mapToScene(event.position().toPoint())
        delta = new_pos - old_pos
        self.translate(delta.x(), delta.y())
        self._mouse_last_press_position = event.position()


    def zoomView(self, event: QWheelEvent) -> None:
        if not len(self.scene().items()) > 0:
            return

        scale_max = 500.0
        scale_min = 0.005
        scale_in_factor = 1.2
        scale_out_factor = 1.0 / scale_in_factor
        scale = self.transform().m11()

        self.updateSceneSize()
        pos = event.position().toPoint()
        old_pos = self.mapToScene(pos)
        angle = event.angleDelta().y()
        factor: float
        if angle > 0.0:
            factor = scale_in_factor
            if scale * factor > scale_max:
                return
        else:
            factor = scale_out_factor
            if scale * factor < scale_min:
                return

        self.setTransformationAnchor(QGraphicsView.ViewportAnchor.NoAnchor)
        self.scale(factor, factor)
        new_pos = self.mapToScene(pos)
        delta = new_pos - old_pos
        self.translate(delta.x(), delta.y())


    def packSelection(self) -> None:
        items = self.scene().selectedItems()
        group = self.scene().createItemGroup(items)
        rect_sizes = []

        for item in items:
            scale = item.scale()
            width = int(item.boundingRect().width() * scale)
            height = int(item.boundingRect().height() * scale)
            rect = (width, height)
            rect_sizes.append(rect)

        bounding_rect = group.boundingRect()
        origin = bounding_rect.topLeft()

        packed_positions = rpack.pack(sizes=rect_sizes)

        for i in range(0, len(items)):
            item = items[i]
            x, y = packed_positions[i]
            pos = origin + QPointF(x, y)
            item.setPos(pos)

        self.scene().destroyItemGroup(group)


    def contextMenuEvent(self, event: QContextMenuEvent) -> None:
        pos = self.mapToGlobal(event.pos())
        system.actions.menu.popup(pos)
        event.accept()


    def updateSceneSize(self) -> None:
        # bypass scrollbar scene size restrictions
        pad = 16000
        widget_rect_in_scene = QRectF(self.mapToScene(-pad, -pad), self.mapToScene(self.rect().bottomRight() + QPoint(pad, pad)))
        new_top_left = QPointF(self.sceneRect().topLeft())
        new_bottom_right = QPointF(self.sceneRect().bottomRight())

        if self.sceneRect().top() > widget_rect_in_scene.top():
            new_top_left.setY(widget_rect_in_scene.top())

        if self.sceneRect().bottom() < widget_rect_in_scene.bottom():
            new_bottom_right.setY(widget_rect_in_scene.bottom())

        if self.sceneRect().left() > widget_rect_in_scene.left():
            new_top_left.setX(widget_rect_in_scene.left())

        if self.sceneRect().right() < widget_rect_in_scene.right():
            new_bottom_right.setX(widget_rect_in_scene.right())

        self.setSceneRect(QRectF(new_top_left, new_bottom_right))


    def onSelectionChanged(self) -> None:
        selected_items = self.scene().selectedItems()
        if len(selected_items) > 0:
            system.actions.setState(True)
        else:
            system.actions.setState(False)


    def rotateSelection(self, event: QMouseEvent) -> None:
        items = self.scene().selectedItems()
        if not len(items) > 0:
            return

        group = self.scene().createItemGroup(items)
        center = group.boundingRect().center()
        group.setTransformOriginPoint(center)

        event_pos = self.mapToScene(event.position().toPoint())
        mid_pos = group.mapToScene(center)
        offset_pos = self.mapToScene(self._mouse_last_press_position.toPoint())

        dx = event_pos - mid_pos
        ax = math.atan2(dx.y(), dx.x())
        dy = offset_pos - mid_pos
        ay = math.atan2(dy.y(), dy.x())
        rotation = math.degrees(ax - ay)

        group.setRotation(group.rotation() + rotation)

        self.scene().destroyItemGroup(group)
        self._mouse_last_press_position = event.position()


    def scaleSelection(self, event: QMouseEvent) -> None:
        items = self.scene().selectedItems()
        if not len(items) > 0:
            return

        # todo: get correct transform scale
        # group = self.scene().createItemGroup(items)
        # center = group.boundingRect().center()
        # group.setTransformOriginPoint(center)

        event_pos = self.mapToScene(event.position().toPoint())
        offset_pos = self.mapToScene(self._mouse_last_press_position.toPoint())
        factor = 0.0005
        scale = (event_pos - offset_pos).x() * factor

        for item in items:
            item_center = item.boundingRect().center()
            # item.setTransformOriginPoint(item_center)
            item.setScale(item.scale() + scale)

        # group.setScale(group.scale() + scale)
        # self.scene().destroyItemGroup(group)
        self._mouse_last_press_position = event.position()

File: src/graphics_view.cpp
// Copyright (C) 2020 viraelin
// License: GPLv3.0-or-later

#include "graphics_view.h"

GraphicsView::GraphicsView(QGraphicsScene *scene, QWidget *parent): QGraphicsView(scene, parent) {
	progress_bar = new QProgressBar(this);
	effect = new QGraphicsColorizeEffect(this);
	network_manager = new QNetworkAccessManager(this);
	image_layer = new QGraphicsRectItem;

	connect(network_manager, &QNetworkAccessManager::finished, this, &GraphicsView::onNetworkReplyFinished);

	setHorizontalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
	setVerticalScrollBarPolicy(Qt::ScrollBarAlwaysOff);
	setAcceptDrops(true);
	setDragMode(DragMode::RubberBandDrag);
	setStyleSheet("background-color: #100111111;");
	setTransformationAnchor(NoAnchor);

	progress_bar->reset();
	progress_bar->setMinimum(0);
	progress_bar->hide();

	QPen pen;
	pen.setStyle(Qt::NoPen);
	QBrush brush = QBrush(QColor("#222222"));
	brush.setStyle(Qt::SolidPattern);
	bounding_rect_item = scene->addRect(QRectF(), pen, brush);
	bounding_rect_item->setZValue(-100);

	effect->setColor(QColor(0, 0, 0));
	effect->setEnabled(false);
	setGraphicsEffect(effect);

	scene->addItem(image_layer);
}

void GraphicsView::projectSave() {
	if (!parentWidget()->isWindowModified()) {
		return;
	}

	QString project_path = current_project_path;
	if (project_path.isEmpty()) {
		const QString path = QStandardPaths::writableLocation(QStandardPaths::DocumentsLocation);
		const QString file_name = QFileDialog::getSaveFileName(
			parentWidget(),
			"Save Project",
			QDir(path).filePath(default_project_name),
			project_filter
		);

		if (file_name.isEmpty()) {
			qDebug() << "Save cancelled or no directory given" << file_name;
			return;
		}
		
		else {
			project_path = file_name;
		}
	}

	QSaveFile file(project_path);
	if (file.open(QIODevice::WriteOnly)) {
		QDataStream out(&file);
		const QList<QGraphicsItem*> items = image_layer->childItems();

		// project settings
		const int version = 100;
		const QPoint view_position = mapToScene(rect().center()).toPoint();
		const qreal zoom = transform().m11();
		const qsizetype item_count = items.size();

		out << version;
		out << view_position;
		out << item_count;
		out << zoom;

		showProgressBar(item_count);

		// save items
		for (int i = 0; i < items.size(); ++i) {
			GraphicsItem *item = static_cast<GraphicsItem*>(items[i]);

			const QPixmap pixmap = item->pixmap();
			QString path = item->data(GraphicsItem::ItemPath).toString();
			const bool is_flipped = item->data(GraphicsItem::ItemIsFlipped).toBool();
			// if zoomed in really far narrowing could matter
			const QPoint position = item->pos().toPoint();
			const qreal scale = item->sceneTransform().m11();
			const qreal z_value = item->zValue();

			if (item->data(GraphicsItem::ItemIsDeleted).toBool()) {
				delete item;
				continue;
			}

			out << path;
			out << pixmap;
			out << is_flipped;
			out << position;
			out << scale;
			out << z_value;

			progress_bar->setValue(i * 100);
			qApp->processEvents();
		}

		progress_bar->hide();
		current_project_path = project_path;
		file.commit();
	}

	else {
		qDebug() << "failed to open save file:" << project_path;
	}
}

void GraphicsView::projectLoad(QString project_path) {
	// default argument workaround
	if (project_path.isEmpty()) {
		project_path = current_project_path;
	}

	QFile file(project_path);
	if (file.open(QIODevice::ReadOnly)) {
		QDataStream in(&file);

		int version;
		QPoint view_position;
		qsizetype item_count;
		qreal zoom;

		in >> version;
		in >> view_position;
		in >> item_count;
		in >> zoom;

		// todo
		if (version != 100) {
			qDebug() << "error: file version invalid:" << version;
			return;
		}

		clearCanvas();
		showProgressBar(item_count);

		const qreal default_scale = 1.0/transform().m11();
		scale(default_scale, default_scale);
		scale(zoom, zoom);

		for (int i = 0; i < item_count; ++i) {
			QString path;
			in >> path;

			if (path.isEmpty()) { continue; }
			// if (!QFile::exists(path)) { continue; }

			QPixmap pixmap;
			bool is_flipped;
			QPoint position;
			qreal scale;
			qreal z_value;

			in >> pixmap;
			in >> is_flipped;
			in >> position;
			in >> scale;
			in >> z_value;

			// todo: merge with createItem or args for GraphicsItem
			GraphicsItem *item = new GraphicsItem(pixmap);
			item->setData(GraphicsItem::ItemPath, path);
			item->setData(GraphicsItem::ItemIsDeleted, false);
			item->setData(GraphicsItem::ItemIsFlipped, false);
			item->setPos(position);
			item->setScale(scale);
			item->setZValue(z_value);
			item->setTransformationMode(transformation_mode);

			if (is_flipped) { item->flip(); };

			item->setParentItem(image_layer);

			progress_bar->setValue(i * 100);
			qApp->processEvents();
		}

		setBoundingRect();
		centerOn(view_position);

		current_project_path = project_path;
		progress_bar->hide();
		file.close();

		QFileInfo info(project_path);
		parentWidget()->setWindowTitle(info.fileName() + "[*]");
	}

	else {
		qDebug() << "failed to open save file:" << project_path;
	}
}

void GraphicsView::mousePressEvent(QMouseEvent *event) {
	if (state == StateDefault) {
		if (event->button() == Qt::LeftButton) {
			QGraphicsView::mousePressEvent(event);
			moving_items = scene()->selectedItems();
			if (moving_items.size() > 0) {
				for (int i = 0; i < moving_items.size(); ++i) {
					QGraphicsItem *item = moving_items[i];
					sendToFront(item);
					moving_items_old_positions.append(item->pos());
				}
			}
		}

		else if (event->button() == Qt::MiddleButton) {
			mouse_last_pan_position = event->position();
			setState(StatePan);
		}
	}

	else if (state == StatePan) {
		if (event->button() == Qt::LeftButton) {
			mouse_last_pan_position = event->position();
		}

		else if (event->button() == Qt::MiddleButton) {
			mouse_last_pan_position = event->position();
		}
	}

	else if (state == StateResize) {
		if (event->button() == Qt::LeftButton) {
			mouse_last_resize_position = event->position();
			const QList<QGraphicsItem*> items = scene()->selectedItems();
			QGraphicsItemGroup *group = scene()->createItemGroup(items);
			resize_origin = group->childrenBoundingRect().center();
			scene()->destroyItemGroup(group);
			setItemParents(items);
		}
	}
}

void GraphicsView::mouseReleaseEvent(QMouseEvent *event) {
	if (state == StateDefault) {
		if (event->button() == Qt::LeftButton) {
			QGraphicsView::mouseReleaseEvent(event);
			if (moving_items.size() > 0) {
				emit itemMoved(moving_items, moving_items_old_positions);
				moving_items.clear();
				moving_items_old_positions.clear();
			}
			setBoundingRect();
		}
	}

	else if (state == StatePan) {
		if (event->button() == Qt::LeftButton) {
			setCursor(Qt::OpenHandCursor);
		}

		else if (event->button() == Qt::MiddleButton) {
			setState(StateDefault);
		}
	}
}

void GraphicsView::mouseMoveEvent(QMouseEvent *event) {
	if (state == StateDefault) {
		QGraphicsView::mouseMoveEvent(event);
		if (moving_items.size() > 0) {
			for (int i = 0; i < moving_items.size(); ++i) {
				QGraphicsItem *item = moving_items[i];
				sendToFront(item);
			}
		}
	}

	else if (state == StatePan) {
		if (event->buttons() == Qt::LeftButton) {
			panView(event);
		}

		else if (event->buttons() == Qt::MiddleButton) {
			panView(event);
		}
	}

	else if (state == StateResize) {
		if (event->buttons() == Qt::LeftButton) {
			resizeSelection(event);
		}
	}
}

void GraphicsView::keyPressEvent(QKeyEvent *event) {
	if (state == StateDefault) {
		if (event->key() == Qt::Key_Space && !event->isAutoRepeat()) {
			setState(StatePan);
		}

		else if (event->key() == Qt::Key_Control && !event->isAutoRepeat()) {
			setState(StateResize);
		}
	}
}

void GraphicsView::keyReleaseEvent(QKeyEvent *event) {
	if (state == StatePan) {
		if (event->key() == Qt::Key_Space && !event->isAutoRepeat()) {
			setState(StateDefault);
		}
	}

	else if (state == StateResize) {
		if (event->key() == Qt::Key_Control && !event->isAutoRepeat()) {
			setState(StateDefault);
		}
	}
}

void GraphicsView::wheelEvent(QWheelEvent *event) {
	zoomView(event);
}

// required
void GraphicsView::dragMoveEvent(QDragMoveEvent *event) {
	Q_UNUSED(event);
	return;
}

void GraphicsView::dragEnterEvent(QDragEnterEvent *event) {
	event->accept();
}

void GraphicsView::dropEvent(QDropEvent *event) {
	mouse_last_drop_position = mapToScene(event->position().toPoint());

	const QList<QUrl> urls = event->mimeData()->urls();
	QList<QString> paths;

	for (int i = 0; i < urls.size(); ++i) {
		const QUrl url = urls[i];
		const QString path = url.toLocalFile();
		if (checkMimeData(path)) {
			paths.append(path);
		}

		else {
			QNetworkRequest request(url);
			QNetworkReply *reply = network_manager->get(request);
			connect(reply, &QNetworkReply::downloadProgress, this, &GraphicsView::onNetworkReplyProgress);
		}
	}

	if (paths.size() > 0) {
		createItems(paths, mouse_last_drop_position);
	}
}

void GraphicsView::setBoundingRect() {
	updateSceneSize();
	QRectF new_rect = image_layer->childrenBoundingRect();
	bounding_rect_item->setRect(new_rect);
}

GraphicsItem *GraphicsView::createItem(const QString &path, const QPointF &pos, const bool is_flipped, const qreal scale, const qreal z_value) {
	GraphicsItem *item = new GraphicsItem(QPixmap(path));
	item->setData(GraphicsItem::ItemPath, path);
	item->setData(GraphicsItem::ItemIsFlipped, false);
	item->setPos(pos);
	item->setScale(scale);
	item->setZValue(z_value);
	item->setTransformationMode(transformation_mode);

	if (is_flipped) {
		item->flip();
	}

	item->setParentItem(image_layer);
	update();
	return item;
}

void GraphicsView::createItems(const QList<QString> &paths, const QPointF &origin) {
	scene()->clearSelection();
	const qsizetype item_count = paths.size();
	showProgressBar(item_count);

	for (int i = 0; i < item_count; ++i) {
		const QString path = paths[i];
		GraphicsItem *item = createItem(path, origin);
		item->setSelected(true);
		progress_bar->setValue(i * 100);
		qApp->processEvents();
	}

	progress_bar->hide();
	packSelection(origin);
}

bool GraphicsView::checkMimeData(const QString &path) {
	const QMimeType mimetype = QMimeDatabase().mimeTypeForFile(path);
	if (mimetype.name().startsWith("image")) {
		return true;
	}

	else {
		return false;
	}
}

void GraphicsView::onNetworkReplyProgress(qint64 bytes_received, qint64 bytes_total) {
	// hack
	const qreal percent = (qreal) bytes_received / (qreal) bytes_total;
	const int value = (percent * 100.0) * 100;
	const int count = 10000;
	showProgressBar(count);
	progress_bar->setValue(value);
}

void GraphicsView::onNetworkReplyFinished(QNetworkReply *reply) {
	// some requests fail: no redirect, no error
	progress_bar->hide();
	const QString temp = QDir::temp().filePath("XXXXXXXX-" + reply->url().fileName());
	QTemporaryFile file(temp);
	qDebug() << file.fileName();
	const QString file_name = file.fileName();

	if (file.open()) {
		qDebug() << file.fileName();
		file.write(reply->readAll());
		file.close();

		if (checkMimeData(file_name)) {
			createItem(file_name, mouse_last_drop_position);
		}
	}
}

void GraphicsView::panView(QMouseEvent *event) {
	updateSceneSize();
	setCursor(Qt::ClosedHandCursor);
	QPointF old_pos = mapToScene(mouse_last_pan_position.toPoint());
	QPointF new_pos = mapToScene(event->position().toPoint());
	QPointF delta = new_pos - old_pos;
	translate(delta.x(), delta.y());
	mouse_last_pan_position = event->position();
}

void GraphicsView::zoomView(QWheelEvent *event) {
	updateSceneSize();
	const QPoint pos = event->position().toPoint();
	const QPointF old_pos = mapToScene(pos);
	const int angle = event->angleDelta().y();
	qreal factor;
	const qreal zoom_in_factor = 1.2;
	const qreal zoom_out_factor = 1.0 / zoom_in_factor;
	factor = (angle > 0.0) ? zoom_in_factor : zoom_out_factor;
	scale(factor, factor);
	const QPointF new_pos = mapToScene(pos);
	const QPointF delta = new_pos - old_pos;
	translate(delta.x(), delta.y());
	setBoundingRect();
}

void GraphicsView::resizeSelection(QMouseEvent *event) {
	const QList<QGraphicsItem*> items = scene()->selectedItems();
	if (items.size() > 0) {
		QGraphicsItemGroup *group = scene()->createItemGroup(items);

		const QPointF epos = event->position();
		const QPointF delta = mouse_last_resize_position - epos;

		const qreal n = -delta.x();
		const int sign = (n > 0) - (n < 0);

		const qreal scene_scale = 1.0/transform().m11();
		const qreal zoom_scale = 0.001 * sign;

		const qreal current_scale = group->scale();
		qreal scale = current_scale + (delta.manhattanLength() * scene_scale) * zoom_scale;
		scale = std::max(scale, 0.01 * scene_scale);
		group->setTransformOriginPoint(resize_origin);
		group->setScale(scale);
		mouse_last_resize_position = epos;
		scene()->destroyItemGroup(group);
		setItemParents(items);
	}
}

void GraphicsView::setState(const State s) {
	if (s == state) {
		return;
	}

	if (s == StateDefault) {
		setCursor(Qt::ArrowCursor);
		setDragMode(DragMode::RubberBandDrag);
	}

	else if (s == StatePan) {
		setCursor(Qt::OpenHandCursor);
		setDragMode(DragMode::NoDrag);
	}

	else if (s == StateResize) {
		setCursor(Qt::SizeHorCursor);
		setDragMode(DragMode::NoDrag);
	}

	state = s;
}

void GraphicsView::setItemParents(const QList<QGraphicsItem*> &items) {
	// hack to try to prevent loss when switching item group
	// use plain rect item as parent because still allows canvas interaction
	for (int i = 0; i < items.size(); ++i) {
		QGraphicsItem *item = items[i];
		item->setParentItem(image_layer);
	}
}

void GraphicsView::packSelection(const QPointF &origin) {
	// simple row packing
	QList<QGraphicsItem*> items = scene()->selectedItems();

	qreal area = 0.0;
	qreal max_width = 0.0;

	for (int i = 0; i < items.size(); ++i) {
		GraphicsItem *item = static_cast<GraphicsItem*>(items[i]);
		qreal scale = item->sceneTransform().m11();
		area += (item->pixmap().width() * scale) * (item->pixmap().height() * scale);
		max_width = std::max(max_width, item->pixmap().width() * scale);
	}

	max_width *= 2.0;

	auto sortByHeight([](QGraphicsItem *_a, QGraphicsItem *_b) {
		GraphicsItem *a = static_cast<GraphicsItem*>(_a);
		GraphicsItem *b = static_cast<GraphicsItem*>(_b);
		const qreal as = a->sceneTransform().m11();
		const qreal bs = b->sceneTransform().m11();
		return (a->pixmap().height() * as) > (b->pixmap().height() * bs);
	});

	std::sort(items.begin(), items.end(), sortByHeight);

	int x = 0;
	int y = 0;
	int largest_height_this_row = 0;

	for (int i = 0; i < items.size(); ++i) {
		GraphicsItem *item = static_cast<GraphicsItem*>(items[i]);
		const QPixmap pixmap = item->pixmap();
		const qreal scale = item->sceneTransform().m11();
		const qreal width = pixmap.width() * scale;
		const qreal height = pixmap.height() * scale;

		// if outside max width, start new row
		if ((x + height) > max_width) {
			x = 0;
			y += largest_height_this_row;
			largest_height_this_row = 0;
		}

		// adjusted position
		const QPointF p(origin.x() + (x - max_width), origin.y() + (y - max_width));
		item->setPos(p);

		x += width;

		if (height > largest_height_this_row) {
			largest_height_this_row = height;
		}
	}
}

void GraphicsView::showProgressBar(const qsizetype &item_count) {
	if (item_count > 0) {
		// set progress bar to center
		// geometry is only correct if window is shown
		const QRect geometry_view = geometry();
		QRect geometry_progress_bar = progress_bar->geometry();
		geometry_progress_bar.moveCenter(geometry_view.center());
		progress_bar->setGeometry(geometry_progress_bar);
		progress_bar->reset();
		progress_bar->show();
		progress_bar->setMaximum((item_count * 100) - 100);
	}
}

void GraphicsView::selectAll() {
	const QList<QGraphicsItem*> items = image_layer->childItems();
	for (int i = 0; i < items.size(); ++i) {
		QGraphicsItem *item = items[i];
		item->setSelected(true);
	}
}

void GraphicsView::clearCanvas() {
	const QList<QGraphicsItem*> current_items = image_layer->childItems();
	for (int i = 0; i < current_items.size(); ++i) {
		QGraphicsItem *item = current_items[i];
		scene()->removeItem(item);
		delete item;
	}
}

void GraphicsView::updateSceneSize() {
	// bypass scrollbar scene size restrictions
	const int pad = 16000;
	const QRectF widget_rect_in_scene(mapToScene(-pad, -pad), mapToScene(rect().bottomRight() + QPoint(pad, pad)));
	QPointF new_top_left(sceneRect().topLeft());
	QPointF new_bottom_right(sceneRect().bottomRight());

	if (sceneRect().top() > widget_rect_in_scene.top()) {
		new_top_left.setY(widget_rect_in_scene.top());
	}

	if (sceneRect().bottom() < widget_rect_in_scene.bottom()) {
		new_bottom_right.setY(widget_rect_in_scene.bottom());
	}

	if (sceneRect().left() > widget_rect_in_scene.left()) {
		new_top_left.setX(widget_rect_in_scene.left());
	}

	if (sceneRect().right() < widget_rect_in_scene.right()) {
		new_bottom_right.setX(widget_rect_in_scene.right());
	}

	setSceneRect(QRectF(new_top_left, new_bottom_right));
}

void GraphicsView::sendToFront(QGraphicsItem *item) {
	qreal max_z = 0;
	const QList<QGraphicsItem*> colliding_items = item->collidingItems(Qt::IntersectsItemBoundingRect);
	for (int i = 0; i < colliding_items.size(); ++i) {
		QGraphicsItem *colliding_item = colliding_items[i];
		if (colliding_item == bounding_rect_item) { continue; }
		max_z = qMax(max_z, colliding_item->zValue());
	}
	item->setZValue(max_z + 1);
}
